File: harness/runner.py
"""Run commands non-interactively with hard timeouts (mitigation #2).

cmd may be a list (preferred - no shell interpretation) or a string.
String commands run with shell=True BY DESIGN: acceptance commands in
tasks.yaml are shell one-liners written by the repo owner, and tasks.yaml
sits under harness/ which the guard blocklist forbids models from editing.
That file is the trust boundary; never pass model-generated text as cmd.
"""
import os
import subprocess

ACTIVE = {}


def run_cmd(cmd, cwd, timeout_s, extra_env=None, tag=None):
    env = {**os.environ, "CI": "true", **(extra_env or {})}
    proc = subprocess.Popen(
        cmd, cwd=cwd, shell=isinstance(cmd, str), env=env, text=True,
        encoding="utf-8", errors="replace",  # tool output is UTF-8, not cp1252
        stdout=subprocess.PIPE, stderr=subprocess.STDOUT,
    )
    if tag:
        ACTIVE.setdefault(tag, set()).add(proc.pid)
    try:
        out, _ = proc.communicate(timeout=timeout_s)
        return proc.returncode, out or ""
    except subprocess.TimeoutExpired:
        # kill the whole tree: a plain kill leaves node children holding ports
        subprocess.run(["taskkill", "/F", "/T", "/PID", str(proc.pid)],
                       capture_output=True)
        try:
            out, _ = proc.communicate(timeout=10)
        except subprocess.TimeoutExpired:
            out = ""
        return 124, (out or "") + "\n[TIMEOUT: process tree killed]"
    finally:
        if tag and tag in ACTIVE:
            ACTIVE[tag].discard(proc.pid)
            if not ACTIVE[tag]:
                del ACTIVE[tag]


def kill_tag(tag):
    for pid in list(ACTIVE.get(tag, ())):
        subprocess.run(["taskkill", "/F", "/T", "/PID", str(pid)],
                       capture_output=True)
    ACTIVE.pop(tag, None)


def tail(text, lines):
    return "\n".join(text.splitlines()[-lines:])

File: src/server/harness/runner.ts
/** Port of backend/harness/runner.py */
import { spawnSync } from "node:child_process";

const ACTIVE = new Map<string, Set<number>>();

export function trackPid(tag: string, pid: number): void {
  const set = ACTIVE.get(tag) ?? new Set<number>();
  set.add(pid);
  ACTIVE.set(tag, set);
}

export function killTag(tag: string): void {
  const pids = ACTIVE.get(tag);
  if (pids) {
    for (const pid of [...pids]) {
      spawnSync("taskkill", ["/F", "/T", "/PID", String(pid)], { encoding: "utf-8" });
    }
  }
  ACTIVE.delete(tag);
}

export function tail(text: string, lines: number): string {
  return text.split(/\r?\n/).slice(-lines).join("\n");
}

export function runCmd(
  cmd: string | string[],
  cwd: string,
  timeoutS: number,
  extraEnv: Record<string, string> = {},
  tag?: string,
): [number, string] {
  const isShell = typeof cmd === "string";
  const r = spawnSync(isShell ? cmd : cmd[0]!, isShell ? [] : cmd.slice(1), {
    cwd,
    shell: isShell,
    encoding: "utf-8",
    env: { ...process.env, CI: "true", ...extraEnv },
    timeout: timeoutS * 1000,
    maxBuffer: 10 * 1024 * 1024,
  });
  if (tag && r.pid) trackPid(tag, r.pid);
  const out = (r.stdout || "") + (r.stderr || "");
  if (r.error?.name === "Error" && String(r.error).includes("ETIMEDOUT")) {
    return [124, `${out}\n[TIMEOUT: process tree killed]`];
  }
  return [r.status ?? 1, out];
}
